common/lib/xmodule/xmodule/x_module.py

```diff
@@ -1177,6 +1177,14 @@ def export_to_xml(self, resource_fs):
         """
         raise NotImplementedError('Modules must implement export_to_xml to enable xml export')
 
+    # =============================== BUILTIN METHODS ==========================
+    def __eq__(self, other):
+        return (hasattr(other, 'scope_ids') and
+                self.scope_ids == other.scope_ids and
+                list(self.fields.keys()) == list(other.fields.keys()) and
+                all(getattr(self, field.name) == getattr(other, field.name)
+                    for field in self.fields.values()))
+
```

Comment:
@bradenmacdonald @ormsbee Haven't figure out fully but without this the following tests fail:

```
Run Tests / cms-unit / cms.djangoapps.contentstore.management.commands.tests.test_delete_orphans.TestDeleteOrphan.test_delete_orphans_published_branch_split
Run Tests / commonlib-unit / xmodule.xmodule.modulestore.tests.test_xml.TestXMLModuleStore.test_dag_course
Run Tests / commonlib-unit / xmodule.xmodule.modulestore.tests.test_xml.TestModuleStoreIgnore.test_dag_course
```

They all fail when checking whether the XBlock instance exists in a list e.g. `self.assertIn(orphan, self.store.get_items(published_branch))`. Any ideas why?

Reply:
I haven't looked into this, but here's my reaction: I don't think there's a good reason to allow this sort of implicit comparison of XBlock/XModule instances. If I were you I'd put this method back where it was before (or factor it out to a `are_blocks_equal` helper method) and then change those tests to _explicitly_ check that the `scope_ids`, `fields.keys()`, and field values are the same.

Reply:
Yeah, given that the other XBlocks don't need it, I agree. I was just wondering why an `in` comparison not work with it. @ormsbee Want to check your recommendation as well?

Reply:
I can certainly imagine scenarios where this sort of comparison could be useful, like if you're comparing something you instantiated in an imported vs. something that's already there, XModules that exist in different branches, roundtrip import/export tests, etc. It would be counterintuitive that two instances created in separate imports would not compare as equal.

Why don't other XBlocks need/use it?

Reply:
@ormsbee @bradenmacdonald I just checked the code around `get_children()` calls in edx-platform and I don't really see any equality comparisons being done between XBlocks. Which is not surprising given that `XBlock` class does not have a `__eq__` implementation and so it wouldn't work.

Looking at the `blame` this may have been done in the early years but dropped at some point when XBlocks started getting written?
https://github.com/edx/edx-platform/blame/master/common/lib/xmodule/xmodule/x_module.py

The two options are
1. Move this to `XModuleDescriptorToXBlockMixin` as done now and we maintain exact behavior for HTML XBlock (though a behavior other code does not seem to depend on).
2. Leave it in `XModuleDescriptor`. That would imply we eventually want to remove this method as further conversions take place?

Which would you rather go with?

Reply:
If it's really not used anywhere, then yeah, let's kill it.

Reply:
Okay, I have moved the `__eq__` method back to where it was.